src/client/ua_client_connect.c

```diff
@@ -1513,6 +1567,145 @@ UA_Client_connectSecureChannel(UA_Client *client, const char *endpointUrl) {
     return client->connectStatus;
 }
 
+static void
+__Client_reverseConnectCallback(UA_ConnectionManager *cm, uintptr_t connectionId,
+                         void *application, void **connectionContext,
+                         UA_ConnectionState state, const UA_KeyValueMap *params,
+                         UA_ByteString msg) {
+    void *reverseConnectIndicator = (void *)(uintptr_t)0xFFFFFFFF;
+
+    UA_Client *client = (UA_Client*)application;
+
+    UA_LOCK(&client->clientMutex);
+
+    if (!client->channel.connectionId && !client->channel.listenerDisconnected) {
+        client->channel.connectionId = connectionId;
+        *connectionContext = reverseConnectIndicator;
+    }
+
+    if (*connectionContext == reverseConnectIndicator && state == UA_CONNECTIONSTATE_CLOSING) {
+        if (client->channel.state == UA_SECURECHANNELSTATE_REVERSE_LISTENING && client->channel.connectionId == connectionId) {
+            client->channel.state = UA_SECURECHANNELSTATE_CLOSED;
+            notifyClientState(client);
+        }
+        UA_UNLOCK(&client->clientMutex);
+        return;
+    }
+
+    if (client->channel.connectionId == connectionId &&  *connectionContext == reverseConnectIndicator) {
+        client->channel.state = UA_SECURECHANNELSTATE_REVERSE_LISTENING;
+        notifyClientState(client);
+    }
+
+    /* This is a connection initiated by a server, disconnect the listener */
+    if (client->channel.connectionId != connectionId && !client->channel.listenerDisconnected) {
+        client->channel.listenerDisconnected = true;
+        cm->closeConnection(cm, client->channel.connectionId);
+    }
+
+    /* First call with the new connection from the server, reset secure channel information */
+    if (client->channel.connectionId != connectionId && *connectionContext == reverseConnectIndicator) {
+        *connectionContext = NULL;
+        client->channel.connectionId = 0;
+        /* Reset to fresh state, the standard network callback will set REVERSE_CONNECTED */
+        client->channel.state = UA_SECURECHANNELSTATE_FRESH;
+    }
+
+    if (*connectionContext != reverseConnectIndicator) {
+        UA_UNLOCK(&client->clientMutex);
+        __Client_networkCallback(cm, connectionId, application, connectionContext, state, params, msg);
+        return;
+    }
+
+    UA_UNLOCK(&client->clientMutex);
+}
+
+UA_StatusCode UA_Client_startListeningForReverseConnect(UA_Client *client, const UA_String *listenHostnames,
+                                                        size_t listenHostnamesLength,
+                                                        UA_UInt16 port) {
+    if (client->channel.state > UA_SECURECHANNELSTATE_FRESH && client->channel.state < UA_SECURECHANNELSTATE_CLOSED)
```

Comment:
Add a log warning here.
Otherwise users without access direct to the source will be confused on what is going on.

Reply:
Done